prov/verbs/src/verbs_info.c

```diff
@@ -199,19 +199,23 @@ int fi_ibv_check_rx_attr(const struct fi_rx_attr *attr,
 			 const struct fi_info *hints,
 			 const struct fi_info *info)
 {
-	/* WARNING: This is not thread safe */
-	uint64_t saved_prov_mode = info->rx_attr->mode;
+	struct fi_info *dup_info;
 	int ret;
 
-	info->rx_attr->mode = (hints->domain_attr &&
-			       hints->domain_attr->cq_data_size) ?
-			info->rx_attr->mode :
-			(info->rx_attr->mode & ~FI_RX_CQ_DATA);
-
-	ret = ofi_check_rx_attr(&fi_ibv_prov, info, attr, hints->mode);
-
-	info->rx_attr->mode = saved_prov_mode;
+	if ((hints->domain_attr && hints->domain_attr->cq_data_size) ||
+	    (hints->rx_attr && hints->rx_attr->mode & FI_RX_CQ_DATA) ||
+	    hints->mode & FI_RX_CQ_DATA) {
```

Comment:
Since we're handling the thread safe case in the else branch, I think the check for cq_data_size should alone be sufficient here as ofi_check_rx_attr would check for FI_RX_CQ_DATA.

Reply:
If we know the hints requested FI_RX_CQ_DATA mode, why do we want to force going through the overhead of the dup of the fi_info just to take it out, shouldn't it just pass?

Reply:
I see. So the app doesn't set cq_data_size but sets FI_RX_CQ_DATA to get to know what cq_data_size it can use.

I was looking at how FI_RX_CQ_DATA is checked twice, once in this condition and once more in ofi_check_rx_attr.